File: kotlin/src/main/kotlin/advertisements/advertisement/infrastructure/persistence/SqLiteAdvertisementRepository.kt
package advertisements.advertisement.infrastructure.persistence

import advertisements.advertisement.domain.Advertisement
import advertisements.advertisement.domain.value_object.ActiveAdvertisements
import advertisements.advertisement.domain.value_object.AdvertisementDate
import advertisements.advertisement.domain.value_object.AdvertisementId
import advertisements.advertisement.domain.value_object.Description
import advertisements.shared.value_object.CivicCenterId
import advertisements.shared.value_object.Email
import advertisements.shared.value_object.Password
import advertisements.shared.value_object.UserId
import advertisements.user.domain.MemberUser
import framework.database.DatabaseConnection
import java.time.LocalDateTime

class SqLiteAdvertisementRepository(private val connection: DatabaseConnection):
    advertisements.advertisement.domain.AdvertisementRepository {
    override fun save(advertisement: Advertisement) {
        val passwordHash = advertisement.password.value()
        connection.execute(
            "INSERT INTO advertisements (id, description, email, password, advertisement_date, civic_center_id, user_id, status, approval_status) VALUES ('" +
                    "${advertisement.id.value()}', '${advertisement.description.value()}', '${advertisement.email.value()}', '$passwordHash', '${advertisement.date.value()}', '${advertisement.civicCenterId.value()}', '${advertisement.memberId.value()}', '${advertisement.status.name}', '${advertisement.approvalStatus.name}') " +
                    "ON CONFLICT(id) DO UPDATE SET description = '${advertisement.description.value()}', email = '${advertisement.email.value()}', password = '${passwordHash}', advertisement_date = '${advertisement.date.value()}', civic_center_id = '${advertisement.civicCenterId.value()}', user_id = '${advertisement.memberId.value()}', status = '${advertisement.status.name}', approval_status = '${advertisement.approvalStatus.name}';"
        )
    }

    override fun findById(id: AdvertisementId): Advertisement? {
        val result = connection.query(
            "SELECT * FROM advertisements WHERE id = '${id.value()}'"
        )

        if (!result.next()) {
            return null
        }

        return Advertisement(
            AdvertisementId(result.getString("id")),
            Description(result.getString("description")),
            Email(result.getString("email")),
            Password.fromEncryptedPassword(result.getString("password")),
            AdvertisementDate(LocalDateTime.parse(result.getString("advertisement_date"))),
            CivicCenterId.create(result.getString("civic_center_id")),
            UserId(result.getString("user_id"))
        )
    }

    override fun activeAdvertisementsByMember(member: MemberUser): ActiveAdvertisements {
        val result = connection.query("SELECT COUNT(*) as active FROM advertisements WHERE user_id = '${member.id().value()}' AND status = 'active'")
        return ActiveAdvertisements.fromInt(result.getInt("active"))
    }

    override fun delete(advertisement: Advertisement) {
        connection.execute("DELETE FROM advertisements WHERE id = '${advertisement.id.value()}'")
    }
}

File: typescript/src/advertisements/advertisement/infrastructure/persistence/SqliteAdvertisementRepository.ts
import { AdvertisementRepository } from '../../domain/AdvertisementRepository';
import { Advertisement } from '../../domain/Advertisement';
import { DatabaseConnection } from '../../../../framework/database/DatabaseConnection';
import {Password} from "../../../shared/domain/value-object/Password";
import {AdvertisementDate} from "../../domain/value-object/AdvertisementDate";
import {Description} from "../../domain/value-object/Description";
import {AdvertisementId} from "../../domain/value-object/AdvertisementId";
import {Email} from "../../../shared/domain/value-object/Email";
import {CivicCenterId} from "../../../shared/domain/value-object/CivicCenterId";
import {UserId} from "../../../shared/domain/value-object/UserId";
import {MemberUser} from "../../../user/domain/MemberUser";
import {ActiveAdvertisements} from "../../domain/value-object/ActiveAdvertisements";

export class SqliteAdvertisementRepository implements AdvertisementRepository {

  constructor(
    private connection: DatabaseConnection) {
  }

  async findById(id: AdvertisementId): Promise<Advertisement | null> {

    const result = await this.connection.query(`SELECT * FROM advertisements WHERE id = ? `, [id.value()])

    if (!result || result.length < 1) {
      return null
    }

    const row = result[0] as any;
    return new Advertisement(
      new AdvertisementId(row.id),
      new Description(row.description),
      new Email(row.email),
      Password.fromEncryptedPassword(row.password),
      new AdvertisementDate(new Date(row.advertisement_date)),
      new CivicCenterId(row.civic_center_id),
      new UserId(row.user_id)
    )
  }

  async activeAdvertisementsByMember(member: MemberUser): Promise<ActiveAdvertisements> {
    const result = await this.connection.query(`SELECT * FROM advertisements WHERE user_id = ? AND status = 'active'`, [member.id().value()])

    return ActiveAdvertisements.fromInt(result.length)
  }

  async save(advertisement: Advertisement): Promise<void> {

    await this.connection.execute(
      `INSERT INTO advertisements (id, description, email, password, advertisement_date, civic_center_id, user_id, status, approval_status) 
      VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?) 
      ON CONFLICT(id) DO UPDATE 
      SET description = excluded.description, password = excluded.password, advertisement_date = excluded.advertisement_date, civic_center_id = excluded.civic_center_id, user_id = excluded.user_id, status = excluded.status, approval_status = excluded.approval_status`, [
      advertisement.id().value(),
      advertisement.description().value(),
      advertisement.email().value(),
      advertisement.password().value(),
      advertisement.date().value().toISOString(),
      advertisement.civicCenterId().value(),
      advertisement.memberId().value(),
      advertisement.status().value(),
      advertisement.approvalStatus().value(),
    ]);
  }

  delete(advertisement: Advertisement): Promise<void> {
      return this.connection.execute(`DELETE FROM advertisements WHERE id = ?`, [advertisement.id().value()])
  }
}
